docs/rfc/0662-collaborative-objects.adoc

```diff
@@ -0,0 +1,778 @@
+counter, timestamp:: as their corresponding integer value
+null:: major type 7, 22
+
+Cursors, undefined values, and constructions which can not be represented as
+definite-length arrays / maps in CBOR are illegal for signed values.
+
+Finally, the signature is computed over the binary CBOR representation
+
+
+=== Strange Perspectives
+
+This model introduces some counter-intuitive properties. For example, I might
+"create an issue" in a repository and anyone who is tracking me would see that
+issue, but people who are tracking the project but don't have me in their
+tracking graph will only see the issue if the maintainer replies to it. It's
+hard to see how you would do things like "link to an issue" under these
+constraints. This is inherent to the network model though, rather than being a
+specific problem of this architecture.
+
+We can work around some of this weirdness using seed nodes. If we consider
+seed nodes 
+
+== APIs
+
+The APIs librad will provide:
+
+* enumerate collaborative objects of a particular type
+* retrieve an object with a particular ID as a JSON representation for
+  applications which only wish to read data
+* retrieve an object with a particular ID as an Automerge document for
+  applications which wish to write data
+* update an object by providing the bytes of an automerge change which updates
+  the document
+* create a new object from a JSON object, a JSON schema, and a type name
+  
+Note that I am referring to "the binary representation of an automerge x" 
+because the automerge API works in terms of binary changes.
+
+This new api will live in a new top level module at
+`librad::collaborative_objects`. An initial sketch looks like this:
+
+[source,rust]
+----
+struct CollaborativeObjectStore {
+    storage: git::storage::Pool,
+    signer: signer::Signer,
+}
+
+enum History {
+    Automerge(Vec<Vec<u8>>)
+}
+
+struct ObjectId(String);
+struct TypeName(String);
+struct Schema(..);
+
+struct CollaborativeObject {
+    typename: TypeName,
+    schema: Schema,
+    id: ObjectId,
+    author: Person, 
+    json: serde_json::Value,
+    history: History, 
+}
+
+struct NewObjectSpec {
+    typename: TypeName,
+    history: History,
+    schema_json: serde_json::Value,
+}
+
+impl CollaborativeObjectStore {
+    fn retrieve_objects(&self, typename: String) -> Result<_, Vec<CollaborativeObject>>
+    fn retrieve_object(&self, typename: String, id: ObjectId) -> Result<_, CollaborativeObject>
+    fn create_object(&self, spec: NewObjectSpec) -> Result<_, CollaborativeObject>
+    fn update_object(&self, id: String, changes: History) -> Result<_, CollaborativeObject>
+}
+----
+
+== Blessed Data Types
+
+This project metadata mechanism is extremely broad, which has a lot of upsides
+but it runs the risk of running into XMPP style extension hell, where every
+peer is running a different set of extensions. It might be worthwhile to bundle
+a few core extensions with librad - issues for example.
+
+
+== Further work
```

Comment:
I think it's also worth calling out that schema migrations are also considered further work